### ISSUE TITLE
Check for browser compatibility in bundle and loader (i.e. don't break IE < 11)

### ISSUE THREAD

**Issue by jhildenbiddle** [webcomponents/webcomponentsjs#1022]
Both `webcomponents-bundle.js` and `webcomponents-loader.js` apply polyfills without testing for browser compatibility. The end result is that legacy browsers (like IE < 11) break as a result. Preferably, a test would be performed that would alllow unsupported browsers to exit gracefully before polyfills are applied.

Old IE. I know. Hear me out... 😉 

I am the author of [css-vars-ponyfill](https://github.com/jhildenbiddle/css-vars-ponyfill), which is a CSS custom property ponyfill that works in legacy browsers as far back as IE9. I recently updated this library to handle shadow DOM `<link>` and `<style>` nodes in response to [this issue](https://github.com/jhildenbiddle/css-vars-ponyfill/issues/24), allowing web component authors to leverage CSS custom properties in legacy browsers starting with IE 11.

I ran into the above issue while writing automated tests for the ponyfill and a dependency ([get-css-data](https://github.com/jhildenbiddle/get-css-data)). Both libraries provide browser compatibility back to IE9, and both provide functionality that can be useful for web component authors (requiring IE 11+). Testing the new shadow DOM functionality added to each library requires loading `webcomponents-bundle.js` for legacy browsers like IE11, but doing so breaks IE 9 and IE 10 which causes other tests to fail in these browsers. I can conditionally run tests, but I cannot conditionally load the webcomponents polyfill. 

As a result of this issue I had to make a choice: remove shadow DOM tests (allowing me to remove webcomponents polyfills), or stop running automated tests in IE9 and IE10. Neither is a good option, but I opted for the latter for the time being.

Fortunately, a basic fix is easy:

```javascript
const isOldIE = window.navigator.appVersion.indexOf("MSIE ") !== -1;

// IE < 11
if (isOldIE) {
  return;
}
```

A more robust solution would use feature detection instead of user agent sniffing, but the above fix feels sufficient until such a solution is requested.

FWIW, I realize that some may argue my issue is with my testing setup not allowing me to conditionally include the polyfill. Working around Karma's shortcomings is one way to work through this, but a well-written polyfill should also "do no harm" when loaded into an environment it cannot support.

Happy to create the PR. Just wanted to run it by the maintainers first. Thx!

**Comment by TimvdLippe** [comment]
Hello there! We provide no support for IE older than IE11 and thus it is expected that things will break when loading this polyfill in IE9/10. Does your test setup also support loading multiple `index.html` to bootstrap your test suites? If so, you can only run the shadow dom tests on IE11+ with the polyfill. All other tests would still run on IE9/10.

**Comment by jhildenbiddle** [comment]
Thanks for the reply, @TimvdLippe.

I understand that this project doesn't support IE < 11. The ask is not to add support for these browsers, but to simply not break them. These are very different things. 

I appreciate the push back on anything related to IE9/10, but I would propose that "doing no harm" by ensuring `webcomponents-bundle.js` and `webcomponents-loader.js` don't break unsupported environments is simply the right thing to do. Assuming the fix in question doesn't present significant technical concerns or challenges, there's no reason for a ponyfill fix some environments while breaking others.

Consider a similar issue filed against one of my own projects: https://github.com/jhildenbiddle/css-vars-ponyfill/issues/16. The author was running css-vars-ponyfill in an SSR environment, and since the ponyfill assumed `window` would be present it would throw an error. I could have responded with "this lib requires a browser / won't work in SSR environments", but I opted to implement [a simple fix](https://github.com/jhildenbiddle/css-vars-ponyfill/commit/ed7864ae2e38db685dc901d691bbcc89883bf2ad#diff-1fdf421c05c1140f6d71444ea2b27638) to avoid forcing others to work around the issue.

Here's another use case that won't work without a fix (because `bundle` and `loader` break IE < 11):

```javascript
if (window.customElements) {
  // Do custom element stuff.
  // Example: Inject <my-custom-select>
}
else {
  // Do legacy stuff
  // Example: Inject <select>
}
```

As far as the testing setup goes, I'm sure I can hack around the issue by forcing tests to run in iframes with HTML fixtures (not how Karma runs by default, so addition configuration and plugins are required) or creating completely separate configuration and test files *just* for web component tests. That feels like a lot of work for what (IMHO) is an issue that should be resolved with the ponyfill.

Thanks for considering, and apologies if it sounds like I'm banging a drum here. My intention is really to 1) selfishly find an easy fix for my projects, and 2) hopefully spare others from having the deal with the same issue. 😄

**Comment by TimvdLippe** [comment]
Shortly discussed this issue with the other maintainers. We can add an extra check for old IE to not run the polyfill. However, my personal hunch is that we should then be consistent and do this for all other browsers (and that by definition is unmaintainable). Not sure what the best middle-ground is here.

**Comment by jhildenbiddle** [comment]
Thanks for taking the time to reconsider and touch base with the other maintainers. Very much appreciated, @TimvdLippe.

I think your person hunch is spot on, btw. Ideally, the polyfill would identify compatible browsers using feature detection instead of user agent sniffing. Typically that would mean detecting the features required by the polyfill to work properly, but this isn't very practicaly given the fact that `webcomponents-bundle.js` and `webcomponents-loader.js` are made up of several other polyfills.

A feature detection "cheat" is to test for a feature that is supported *only* by the target platforms, regardless of whether the polyfill uses that feature. The [comparison feature of caniuse.com](https://caniuse.com/#comparison) is a handy tool for this. I didn't find a perfect match, but here are two good candidates:

- [CSS font-feature-settings](https://caniuse.com/#feat=font-feature)
   Supported by IE10, so a separate test for IE < 11 is required.

   ```javascript
   // Detect CSS font-feature-settings & IE < 11
   var isBrowser   = this.window === this;
   var isOldIE     = isBrowser && window.navigator.appVersion.indexOf("MSIE ") !== -1;
   var isSupported = isBrowser && 'fontFeatureSettings' in document.body.style;

   if (isOldIE || !isSupported) {
     return;
   }
   ```

   FWIW, feature support similar to CSS `supports()` is available for [String.prototype.includes](https://caniuse.com/#feat=es6-string-includes), [ES6 Number](https://caniuse.com/#feat=es6-number), and [ES6 classes](https://caniuse.com/#feat=es6-class), so any of these could be used.

- [CSS.supports() API](https://caniuse.com/#feat=css-supports-api)
   Not supported by IE11, so a separate test for IE 11 is required.

   ```javascript
   // Detect CSS.supports() API & IE 11
   var isBrowser   = this.window === this;
   var isIE11      = isBrowser && !!window.MSInputMethodContext && !!document.documentMode;
   var isSupported = isBrowser && (window.CSS || {}).supports !== undefined;

   if (!isIE11 && !isSupported) {
     return;
   }
   ```

Both of the above solutions test as follows:

- **Not Supported:** Chrome 47, Firefox 33, IE 10, Safari 8
- **Supported:** Chrome 48, Firefox 34, IE 11, Safari 9

This seems a good middle ground to me, and a better solution than my initial "oldIE" proposal.

**Comment by jhildenbiddle** [comment]
Good to know, @jsilvermist. Perhaps the following is a better solution?

```javascript
var isBrowser = typeof window !== 'undefined';
```

**Comment by web-padawan** [comment]
Consider the fact that polyfills are being tested also agains Chrome 41 because of the Googlebot still using it. Wondering it it would be ever upgraded, hopefully next year 🤔

**Comment by TimvdLippe** [comment]
@jhildenbiddle Rather than targeting a browser, I would like to know what specific properties have to hold for the polyfill to go wrong. E.g. if we can detect "Creating an element X with Y" breaks in a browser, and that property is a prerequisite for the polyfill, we should test for that.

For example, see https://github.com/webcomponents/custom-elements/blob/e852c5006497b22e51e6f40d4fbfedb032c64533/tests/html/Element/insertAdjacentHTML.html#L171-L181 where we explicitly test for a prerequisite. If the property holds, we do NOT run the test. If you would be able to figure out why IE9/10 fail and be able to write a property check for that, I think that is :100:

**Comment by jhildenbiddle** [comment]
@TimvdLippe - The approach you're proposing makes sense. It's the "correct" way use feature detection to test if a polyfill *can* execute code (i.e. all requirements have been met) and *should* execute code (i.e. when the environment needs to be polyfilled). It's also a significantly larger effort that—in this particular scenario—doesn't yield benefits that warrant the effort (IMHO). Hence the two simplified approaches: skipping IE < 11 via user agent sniffing or skipping several legacy browsers using a feature detection "cheat".

I do appreciate the desire to do things to right way. Is the option to implement an "OldIE" test still on the table? This would solve the immediate issue for IE < 11 and get the "only proceed if X" logic in place (which could later be extended with additional tests if necessary).

Thanks for the follow-up and continued discussion, and no worries if you and/or the team don't have the cycles to spare.

**Comment by TimvdLippe** [comment]
> doesn't yield benefits that warrant the effort

The benefit is that we do not discriminate any browsers based on UA sniffing. Instead, we check for explicit support for a certain feature and filter on that. Thus, it is a maintainable solution and a solution I would accept. The only thing is: I currently do not know what IE9 is lacking/breaking that is preventing our polyfills from working. If we can figure out what is actually breaking, we can filter on that and make it work.

**Comment by justinfagnani** [comment]
@jhildenbiddle what actually breaks in unsupported browsers?

**Comment by jhildenbiddle** [comment]
@justinfagnani -- A barebones HTML page loading only `webcomponents-bundle.js` via CDN displays the following message in the IE9/10's console:

```
SCRIPT5009: 'MutationObserver' is undefined 
webcomponents-bundle.js, line 73 character 48
```

Adding a `MutaionObserver` polyfill then produces the following:

```
SCRIPT5009: 'Set' is undefined 
webcomponents-bundle.js, line 136 character 415
```

Adding a `Set` polyfill then produces the following in IE9 (not IE10):

```
SCRIPT5009: 'requestAnimationFrame' is undefined 
webcomponents-bundle.js, line 189 character 278
```

Adding a `requestAnimationFrame` polyfill for IE9 then leaves us with the following for both IE9/10:

```
SCRIPT438: Object doesn't support property or method 'composedPath' 
index.html, line 1 character 1
```

Attempts to ponyfill `Event.composedPath()` were unsuccessful, so the adventure comes to an end.

FWIW, I realize we've spent a decent amount of time discussing issues related to browsers few people care about anymore. My hope was that this would be an easy and uncontroversial change that provided value (even if only for a small group of users), but I understand this may not be the case for a variety of reasons. I'm happy to continue the discussion and help with effort, but I also fine closing the issue if the consensus among the maintainers is that this is not something they want to pursue.

Thanks for the follow-up @justinfagnani and the for the initial discussion @TimvdLippe. Much appreciated.

**Comment by TimvdLippe** [comment]
@jhildenbiddle These results are very interesting! It is indeed expected you polyfill these browser features. However, the `composedPath` issue is interestintg. We polyfill `composedPath` in https://github.com/webcomponents/shadydom/blob/e30766a62aa193860a86b50169b127588d41d976/src/patch-events.js#L153-L158

This means that we are probably missing a prototype override on older IE at https://github.com/webcomponents/shadydom/blob/e30766a62aa193860a86b50169b127588d41d976/src/patch-events.js#L538-L561

I am more than happy to fix the issue in our polyfill so that IE10 "just" works. Would you have enough pointers to potentially tackle this issue?

**Comment by justinfagnani** [comment]
@jhildenbiddle thanks for the. The reason I ask "what breaks?" is because from the original comment it sounded like the polyfill might be breaking something else on the page.

Is the breakage you're talking about just the internal breakage of the polyfills and the errors they report? Do those errors cause the rest of the page to not function?

That would be a big motivator for selectively applying the polyfill. If the check is just to get rid of exceptions internal to the polyfill that don't break code later in the page, it seems less of a critical issue to me.

**Comment by jhildenbiddle** [comment]
After digging in a little deeper, it looks like the errors I encountered were not caused by the webcomponentsjs polyfill. 🤦‍♂️ 

There were three separate errors I ran into:

1. Errors reported in the console in legacy browsers (IE < 11) after adding the webcomponentsjs polyfill. These appear to be harmless, although I am not familiar enough with the polyfill to know if modifications are happening under the hood that may cause problems elsewhere. I do know that the `customElements` is not added to the `window` object in legacy browsers which is good.

   Here's an example:

   ```javacsript
   SCRIPT5009: 'MutationObserver' is undefined 
   ```

1. Errors introduced by adding additional polyfills (courtesy of polyfill.io) to try and work around webcomponentsjs issues in IE < 11. I don't know what the issue is (I didn't spent any time investigating) but adding polyfill.io polyfills resulted in a previously working Mocha+Chai test using `to.include.members` to fail with the following error:

   ```javascript
   SCRIPT438: Object doesn't support property or method 'members'
   ```

   My (incorrect) assumption was that this error was caused by the webcomponentsjs polyfill. Changing the test to use `to.have.members` fixed the issue, so this appears to be polyfill.io and/or Chai issue.

1. Errors introduced by loading web components defined using ES6 classes and transpiled to ES5 in legacy browsers (IE < 11). For example, transpiling the following code:

   ```javascript
   class TestComponent extends HTMLElement {
     // ...
   }
   
   if ('customElements' in window) {
       window.customElements.define('test-component', TestComponent);
   }
   ```

   Produces the following error:

   ```
   SCRIPT438: Object doesn't support property or method 'setPrototypeOf'
   ```

   I assumed (also incorrectly) that the `if` statement would prevent errors by only registering the custom element on compatible browsers, so the error I was seeing must be caused by the webcomponetsjs polyfill. Apparently simply loading the transpiled ES6 class code in a legacy browser is enough to throw an error though. I didn't realize this would be an issue until I did some more research.

Whew.

Long story short, I don't think any changes need to be made to the webcomponentsjs polyfill.

Thanks to both @TimvdLippe and @justinfagnani for sticking with the thread and your willingness to resolve the issue. Much appreciated.

**Comment by TimvdLippe** [comment]
@jhildenbiddle That is unexpected, yet an awesome result! Sorry for being "harsh" and asking for additional details, but in the end we were able to determine the root cause, which I am really happy about. Sorry for all the errors you were encountering that were complicating the story. Polyfilling is hard :cry: 

If you do encounter any issues later on the older browsers, I am still interested in fixing them, as long as we can determine what requires patching. For now, I am closing this issue as the original issue has been resolved. However, feel free to reopen for the above if any future issues arise. Thanks for reaching out to us!

**Comment by jhildenbiddle** [comment]
@TimvdLippe -- No need to apologize. Polyfilling is hard, but not as hard as maintaining a popular open source project and trying to make everyone happy. 😉 

Thanks again, and keep up the great work!
